File: api_server/tools/database_tool.py
"""SQLite Database Tool — create, query, and manage local SQLite databases."""
import json
import os
import re
from typing import Any, Dict, List, Optional

import aiosqlite

from .types import ToolDef, ToolResult, ToolContext

DB_DIR = os.path.join(os.path.expanduser("~"), ".claude", "databases")
MAX_OUTPUT_SIZE = 50000
MAX_ROW_LIMIT = 1000


def _ensure_dir() -> None:
    if not os.path.exists(DB_DIR):
        os.makedirs(DB_DIR, exist_ok=True)


def _get_db_path(name: str) -> str:
    safe = re.sub(r"[^a-zA-Z0-9_-]", "_", name)
    return os.path.join(DB_DIR, f"{safe}.db")


async def _open_db(name: str) -> aiosqlite.Connection:
    _ensure_dir()
    db_path = _get_db_path(name)
    db = await aiosqlite.connect(db_path)
    db.row_factory = aiosqlite.Row
    return db


async def _list_databases_impl() -> str:
    _ensure_dir()
    files = [f for f in os.listdir(DB_DIR) if f.endswith(".db")]
    dbs = []
    for f in files:
        stat = os.stat(os.path.join(DB_DIR, f))
        dbs.append({
            "name": f[:-3],
            "size": f"{(stat.st_size / 1024):.1f} KB",
            "modified": stat.st_mtime,
        })
    return json.dumps({"databases": dbs, "count": len(dbs), "directory": DB_DIR}, indent=2)


async def _list_tables_impl(db: aiosqlite.Connection, db_name: str) -> str:
    cursor = await db.execute(
        "SELECT name, sql FROM sqlite_master WHERE type='table' AND name NOT LIKE 'sqlite_%' ORDER BY name"
    )
    rows = await cursor.fetchall()
    tables = [{"name": row["name"], "sql": row["sql"]} for row in rows]
    return json.dumps({"database": db_name, "tables": tables}, indent=2)


async def _describe_table_impl(db: aiosqlite.Connection, db_name: str, table: str) -> str:
    cursor = await db.execute(f"PRAGMA table_info(\"{table}\")")
    columns = await cursor.fetchall()
    
    count_cursor = await db.execute(f'SELECT COUNT(*) as count FROM "{table}"')
    count_row = await count_cursor.fetchone()
    count = count_row["count"] if count_row else 0
    
    return json.dumps({
        "database": db_name,
        "table": table,
        "columns": [dict(row) for row in columns],
        "rowCount": count,
    }, indent=2)


async def _create_table_impl(db: aiosqlite.Connection, sql: str) -> str:
    await db.execute(sql)
    await db.commit()
    return "Table created successfully."


async def _query_impl(db: aiosqlite.Connection, db_name: str, sql: str, params: Optional[List] = None) -> str:
    if params:
        cursor = await db.execute(sql, params)
    else:
        cursor = await db.execute(sql)
    
    rows = await cursor.fetchall()
    
    if len(rows) > MAX_ROW_LIMIT:
        rows = rows[:MAX_ROW_LIMIT]
        truncated = True
    else:
        truncated = False
    
    results = [dict(row) for row in rows]
    output = json.dumps({
        "database": db_name,
        "results": results,
        "rowCount": len(results),
        "truncated": truncated,
    }, indent=2, default=str)
    
    if len(output) > MAX_OUTPUT_SIZE:
        output = output[:MAX_OUTPUT_SIZE] + "\n... (truncated)"
    
    return output


async def _execute_impl(db: aiosqlite.Connection, db_name: str, sql: str, params: Optional[List] = None) -> str:
    if params:
        cursor = await db.execute(sql, params)
    else:
        cursor = await db.execute(sql)
    
    await db.commit()
    changes = db.total_changes
    last_id = cursor.lastrowid if cursor.lastrowid else 0
    
    return json.dumps({
        "database": db_name,
        "changes": changes,
        "lastInsertRowid": last_id,
        "message": f"Executed successfully. {changes} row(s) affected.",
    }, indent=2)


async def _import_data_impl(db: aiosqlite.Connection, db_name: str, table: str, columns: List[str], rows: List[List]) -> str:
    placeholders = ", ".join(["?" for _ in columns])
    column_names = ", ".join([f'"{c}"' for c in columns])
    sql = f'INSERT INTO "{table}" ({column_names}) VALUES ({placeholders})'
    
    await db.execute("BEGIN TRANSACTION")
    try:
        count = 0
        for row in rows:
            await db.execute(sql, row)
            count += 1
        await db.commit()
        return json.dumps({
            "database": db_name,
            "table": table,
            "rowsInserted": count,
            "message": f"Imported {count} rows.",
        }, indent=2)
    except Exception:
        await db.execute("ROLLBACK")
        raise


async def _export_csv_impl(db: aiosqlite.Connection, sql: Optional[str], table: Optional[str]) -> str:
    if sql:
        query = sql
    elif table:
        query = f'SELECT * FROM "{table}"'
    else:
        return json.dumps({"error": "Either sql or table is required"}, indent=2)
    
    cursor = await db.execute(query)
    rows = await cursor.fetchall()
    
    if len(rows) == 0:
        return "No data to export."
    
    headers = list(rows[0].keys())
    csv_lines = [",".join(headers)]
    
    for row in rows:
        values = []
        for h in headers:
            val = row[h]
            if val is None:
                val_str = ""
            else:
                val_str = str(val)
                if "," in val_str or '"' in val_str or "\n" in val_str:
                    val_str = '"' + val_str.replace('"', '""') + '"'
            values.append(val_str)
        csv_lines.append(",".join(values))
    
    csv = "\n".join(csv_lines)
    
    if len(csv) > MAX_OUTPUT_SIZE:
        csv = csv[:MAX_OUTPUT_SIZE] + "\n... (truncated)"
    
    return csv


async def _drop_table_impl(db: aiosqlite.Connection, table: str) -> str:
    await db.execute(f'DROP TABLE IF EXISTS "{table}"')
    await db.commit()
    return f'Table "{table}" dropped successfully.'


DATABASE_TOOL = ToolDef(
    name="database",
    description="""SQLite database tool. Create, query, and manage local SQLite databases.

Actions:
- "create_table": Create a new table. Params: database (string), sql (CREATE TABLE statement)
- "query": Run a SELECT query. Params: database (string), sql (SELECT statement), params? (array of values for ? placeholders)
- "execute": Run INSERT/UPDATE/DELETE/ALTER. Params: database (string), sql (statement), params? (array of values for ? placeholders)
- "list_databases": List all databases
- "list_tables": List tables in a database. Params: database (string)
- "describe_table": Get table schema. Params: database (string), table (string)
- "export_csv": Export a table or query result as CSV. Params: database (string), sql? (SELECT query), table? (table name if no sql)
- "drop_table": Drop a table. Params: database (string), table (string)
- "import_data": Insert multiple rows. Params: database (string), table (string), columns (string[]), rows (any[][])

Databases are stored in ~/.claude/databases/ as .db files.""",
    input_schema={
        "type": "object",
        "properties": {
            "action": {
                "type": "string",
                "enum": ["create_table", "query", "execute", "list_databases", "list_tables", "describe_table", "export_csv", "drop_table", "import_data"],
                "description": "The action to perform",
            },
            "database": {"type": "string", "description": "Database name (without .db extension)"},
            "sql": {"type": "string", "description": "SQL statement to execute"},
            "table": {"type": "string", "description": "Table name (for describe_table, export_csv, drop_table, import_data)"},
            "params": {"type": "array", "description": "Parameters for parameterized queries (? placeholders)"},
            "columns": {"type": "array", "items": {"type": "string"}, "description": "Column names for import_data"},
            "rows": {"type": "array", "items": {"type": "array"}, "description": "Row data for import_data"},
        },
        "required": ["action"],
    },
    is_read_only=False,
    risk_level="medium",
    execute=lambda args, ctx: _database_execute(args, ctx),
)


async def _database_execute(args: Dict[str, Any], ctx: ToolContext) -> ToolResult:
    tool_call_id = getattr(ctx, "tool_call_id", "") or ""
    
    action = args.get("action")
    db_name = args.get("database")
    sql = args.get("sql")
    table = args.get("table")
    params = args.get("params")
    columns = args.get("columns")
    rows = args.get("rows")
    
    try:
        if action == "list_databases":
            output = await _list_databases_impl()
            return ToolResult(tool_call_id=tool_call_id, output=output, is_error=False)
        
        if action == "list_tables":
            if not db_name:
                return ToolResult(tool_call_id=tool_call_id, output="Error: database name is required", is_error=True)
            db = await _open_db(db_name)
            try:
                output = await _list_tables_impl(db, db_name)
                return ToolResult(tool_call_id=tool_call_id, output=output, is_error=False)
            finally:
                await db.close()
        
        if action == "describe_table":
            if not db_name or not table:
                return ToolResult(tool_call_id=tool_call_id, output="Error: database and table are required", is_error=True)
            db = await _open_db(db_name)
            try:
                output = await _describe_table_impl(db, db_name, table)
                return ToolResult(tool_call_id=tool_call_id, output=output, is_error=False)
            finally:
                await db.close()
        
        if action == "create_table":
            if not db_name or not sql:
                return ToolResult(tool_call_id=tool_call_id, output="Error: database and sql are required", is_error=True)
            db = await _open_db(db_name)
            try:
                output = await _create_table_impl(db, sql)
                return ToolResult(tool_call_id=tool_call_id, output=output, is_error=False)
            finally:
                await db.close()
        
        if action == "query":
            if not db_name or not sql:
                return ToolResult(tool_call_id=tool_call_id, output="Error: database and sql are required", is_error=True)
            sql_upper = sql.strip().upper()
            if not (sql_upper.startswith("SELECT") or sql_upper.startswith("WITH") or sql_upper.startswith("PRAGMA")):
                return ToolResult(
                    tool_call_id=tool_call_id,
                    output="Error: query action only supports SELECT/WITH/PRAGMA statements. Use 'execute' for modifications.",
                    is_error=True,
                )
            db = await _open_db(db_name)
            try:
                output = await _query_impl(db, db_name, sql, params)
                return ToolResult(tool_call_id=tool_call_id, output=output, is_error=False)
            finally:
                await db.close()
        
        if action == "execute":
            if not db_name or not sql:
                return ToolResult(tool_call_id=tool_call_id, output="Error: database and sql are required", is_error=True)
            db = await _open_db(db_name)
            try:
                output = await _execute_impl(db, db_name, sql, params)
                return ToolResult(tool_call_id=tool_call_id, output=output, is_error=False)
            finally:
                await db.close()
        
        if action == "import_data":
            if not db_name or not table or not columns or not rows:
                return ToolResult(tool_call_id=tool_call_id, output="Error: database, table, columns, and rows are required", is_error=True)
            db = await _open_db(db_name)
            try:
                output = await _import_data_impl(db, db_name, table, columns, rows)
                return ToolResult(tool_call_id=tool_call_id, output=output, is_error=False)
            finally:
                await db.close()
        
        if action == "export_csv":
            if not db_name:
                return ToolResult(tool_call_id=tool_call_id, output="Error: database is required", is_error=True)
            db = await _open_db(db_name)
            try:
                output = await _export_csv_impl(db, sql, table)
                return ToolResult(tool_call_id=tool_call_id, output=output, is_error=False)
            finally:
                await db.close()
        
        if action == "drop_table":
            if not db_name or not table:
                return ToolResult(tool_call_id=tool_call_id, output="Error: database and table are required", is_error=True)
            db = await _open_db(db_name)
            try:
                output = await _drop_table_impl(db, table)
                return ToolResult(tool_call_id=tool_call_id, output=output, is_error=False)
            finally:
                await db.close()
        
        return ToolResult(tool_call_id=tool_call_id, output=f"Unknown action: {action}", is_error=True)
    
    except Exception as err:
        return ToolResult(tool_call_id=tool_call_id, output=f"Database error: {err}", is_error=True)


__all__ = ["DATABASE_TOOL"]

File: webui/tools/databaseTool.ts
/**
 * SQLite Database Tool — AI can create, query, and manage SQLite databases.
 * Databases stored in ~/.claude/databases/
 */
import * as path from 'path'
import * as fs from 'fs'
import { Database } from 'bun:sqlite'
import type { ToolDef } from './types.js'

const DB_DIR = path.join(process.env.HOME || process.env.USERPROFILE || '.', '.claude', 'databases')

function ensureDir() {
  if (!fs.existsSync(DB_DIR)) fs.mkdirSync(DB_DIR, { recursive: true })
}

function getDbPath(name: string): string {
  // Sanitize name
  const safe = name.replace(/[^a-zA-Z0-9_-]/g, '_')
  return path.join(DB_DIR, `${safe}.db`)
}

function openDb(name: string): Database {
  ensureDir()
  const dbPath = getDbPath(name)
  return new Database(dbPath)
}

export const databaseTool: ToolDef = {
  name: 'database',
  description: `SQLite database tool. Create, query, and manage local SQLite databases.

Actions:
- "create_table": Create a new table. Params: database (string), sql (CREATE TABLE statement)
- "query": Run a SELECT query. Params: database (string), sql (SELECT statement)
- "execute": Run INSERT/UPDATE/DELETE/ALTER. Params: database (string), sql (statement), params? (array of values for ? placeholders)
- "list_databases": List all databases
- "list_tables": List tables in a database. Params: database (string)
- "describe_table": Get table schema. Params: database (string), table (string)
- "export_csv": Export a table or query result as CSV. Params: database (string), sql? (SELECT query, default all rows), table? (table name if no sql)
- "drop_table": Drop a table. Params: database (string), table (string)
- "import_data": Insert multiple rows. Params: database (string), table (string), columns (string[]), rows (any[][])

Databases are stored in ~/.claude/databases/ as .db files.`,
  inputSchema: {
    type: 'object',
    properties: {
      action: {
        type: 'string',
        enum: ['create_table', 'query', 'execute', 'list_databases', 'list_tables', 'describe_table', 'export_csv', 'drop_table', 'import_data'],
        description: 'The action to perform',
      },
      database: { type: 'string', description: 'Database name (without .db extension)' },
      sql: { type: 'string', description: 'SQL statement to execute' },
      table: { type: 'string', description: 'Table name (for describe_table, export_csv, drop_table, import_data)' },
      params: { type: 'array', description: 'Parameters for parameterized queries (? placeholders)' },
      columns: { type: 'array', items: { type: 'string' }, description: 'Column names for import_data' },
      rows: { type: 'array', items: { type: 'array' }, description: 'Row data for import_data' },
    },
    required: ['action'],
  },
  isReadOnly: false,
  riskLevel: 'medium',
  execute: async (input: Record<string, unknown>) => {
    const action = input.action as string
    const dbName = input.database as string | undefined
    const sql = input.sql as string | undefined
    const table = input.table as string | undefined
    const params = input.params as unknown[] | undefined
    const columns = input.columns as string[] | undefined
    const rows = input.rows as unknown[][] | undefined

    try {
      switch (action) {
        case 'list_databases': {
          ensureDir()
          const files = fs.readdirSync(DB_DIR).filter(f => f.endsWith('.db'))
          const dbs = files.map(f => {
            const stat = fs.statSync(path.join(DB_DIR, f))
            return {
              name: f.replace('.db', ''),
              size: `${(stat.size / 1024).toFixed(1)} KB`,
              modified: new Date(stat.mtimeMs).toISOString(),
            }
          })
          return { output: JSON.stringify({ databases: dbs, count: dbs.length, directory: DB_DIR }, null, 2), isError: false }
        }

        case 'list_tables': {
          if (!dbName) return { output: 'Error: database name is required', isError: true }
          const db = openDb(dbName)
          try {
            const tables = db.query("SELECT name, sql FROM sqlite_master WHERE type='table' AND name NOT LIKE 'sqlite_%' ORDER BY name").all()
            return { output: JSON.stringify({ database: dbName, tables }, null, 2), isError: false }
          } finally {
            db.close()
          }
        }

        case 'describe_table': {
          if (!dbName || !table) return { output: 'Error: database and table are required', isError: true }
          const db = openDb(dbName)
          try {
            const info = db.query(`PRAGMA table_info("${table}")`).all()
            const count = db.query(`SELECT COUNT(*) as count FROM "${table}"`).get() as Record<string, number>
            return { output: JSON.stringify({ database: dbName, table, columns: info, rowCount: count?.count || 0 }, null, 2), isError: false }
          } finally {
            db.close()
          }
        }

        case 'create_table': {
          if (!dbName || !sql) return { output: 'Error: database and sql are required', isError: true }
          const db = openDb(dbName)
          try {
            db.exec(sql)
            return { output: `Table created successfully in database "${dbName}".`, isError: false }
          } finally {
            db.close()
          }
        }

        case 'query': {
          if (!dbName || !sql) return { output: 'Error: database and sql are required', isError: true }
          if (!sql.trim().toUpperCase().startsWith('SELECT') && !sql.trim().toUpperCase().startsWith('WITH') && !sql.trim().toUpperCase().startsWith('PRAGMA')) {
            return { output: 'Error: query action only supports SELECT/WITH/PRAGMA statements. Use "execute" for modifications.', isError: true }
          }
          const db = openDb(dbName)
          try {
            const results = params ? db.query(sql).all(...params) : db.query(sql).all()
            const outputStr = JSON.stringify({ database: dbName, results, rowCount: results.length }, null, 2)
            return { output: outputStr.length > 50000 ? outputStr.slice(0, 50000) + '\n... (truncated)' : outputStr, isError: false }
          } finally {
            db.close()
          }
        }

        case 'execute': {
          if (!dbName || !sql) return { output: 'Error: database and sql are required', isError: true }
          const db = openDb(dbName)
          try {
            const result = params ? db.run(sql, ...params) : db.run(sql)
            return {
              output: JSON.stringify({
                database: dbName,
                changes: result.changes,
                lastInsertRowid: Number(result.lastInsertRowid),
                message: `Executed successfully. ${result.changes} row(s) affected.`,
              }, null, 2),
              isError: false,
            }
          } finally {
            db.close()
          }
        }

        case 'import_data': {
          if (!dbName || !table || !columns || !rows) return { output: 'Error: database, table, columns, and rows are required', isError: true }
          const db = openDb(dbName)
          try {
            const placeholders = columns.map(() => '?').join(', ')
            const stmt = db.prepare(`INSERT INTO "${table}" (${columns.map(c => `"${c}"`).join(', ')}) VALUES (${placeholders})`)
            let count = 0
            db.exec('BEGIN TRANSACTION')
            for (const row of rows) {
              stmt.run(...row)
              count++
            }
            db.exec('COMMIT')
            return { output: JSON.stringify({ database: dbName, table, rowsInserted: count, message: `Imported ${count} rows.` }, null, 2), isError: false }
          } catch (err) {
            db.exec('ROLLBACK')
            throw err
          } finally {
            db.close()
          }
        }

        case 'export_csv': {
          if (!dbName) return { output: 'Error: database is required', isError: true }
          const db = openDb(dbName)
          try {
            const query = sql || (table ? `SELECT * FROM "${table}"` : null)
            if (!query) return { output: 'Error: either sql or table is required', isError: true }
            const results = db.query(query).all() as Record<string, unknown>[]
            if (results.length === 0) return { output: 'No data to export.', isError: false }
            const headers = Object.keys(results[0]!)
            const csvLines = [headers.join(',')]
            for (const row of results) {
              csvLines.push(headers.map(h => {
                const val = row[h]
                const str = val === null ? '' : String(val)
                return str.includes(',') || str.includes('"') || str.includes('\n') ? `"${str.replace(/"/g, '""')}"` : str
              }).join(','))
            }
            const csv = csvLines.join('\n')
            return { output: csv.length > 50000 ? csv.slice(0, 50000) + '\n... (truncated)' : csv, isError: false }
          } finally {
            db.close()
          }
        }

        case 'drop_table': {
          if (!dbName || !table) return { output: 'Error: database and table are required', isError: true }
          const db = openDb(dbName)
          try {
            db.exec(`DROP TABLE IF EXISTS "${table}"`)
            return { output: `Table "${table}" dropped from database "${dbName}".`, isError: false }
          } finally {
            db.close()
          }
        }

        default:
          return { output: `Unknown action: ${action}`, isError: true }
      }
    } catch (err: unknown) {
      return { output: `Database error: ${err instanceof Error ? err.message : String(err)}`, isError: true }
    }
  },
}
